Mathlib/NumberTheory/SiegelsLemma.lean

```diff
@@ -0,0 +1,118 @@
+/-
+Copyright (c) 2024 Fabrizio Barroero. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fabrizio Barroero, Laura Capuano, Amos Turchet
+-/
+import Mathlib.Analysis.Matrix
+import Mathlib.Data.Matrix.Basic
+import Mathlib.Data.Pi.Interval
+
+/-!
+# Siegel's Lemma, first version for ℤ
+
+In this file we introduce and prove Siegel's Lemma in its most basic version. This is a fundamental
+tool in diophantine approximation and transcendency and says that there exists a "small" integral
+non-zero solution of a non-trivial overdetermined system of linear equations with integer
+coefficients.
+
+## Main results
+
+- `siegels_lemma`: Given a non-singular `m × n` matrix `A` with `m < n` the linear system it
+determines has a non-zero integer solution `t` with `‖t‖ ≤ ((n * ‖A‖) ^ ((m : ℝ) / (n - m)))`
+
+## Notation
+
+ - `‖_‖ ` : Matrix.seminormedAddCommGroup is the sup norm, the maximum of the absolute values of
+ the entries of the matrix
+
+## References
+
+See [M. Hindry and J. Silverman, Diophantine Geometry: an Introduction] .
+-/
+
+/- We set ‖⬝‖ to be Matrix.seminormedAddCommGroup  -/
```

Comment:
Most of the results below have nothing special to do with ℕ and ℝ, but with a general cast function. Also, they are part of the matrix library, so they should be moved to a file where the sup-norm is defined.

Reply:
I just got rid of cast_sup_eq_sup_cast_Nat_NNReal because I could just call comp_sup_eq_sup_comp_of_is_total where needed. I will then move these lemmas in the sup-norm folder.